File: src/services/reportsRoutingService.ts
import { auth, db } from '../config/firebase';
import { doc, getDoc } from 'firebase/firestore';

/**
 * Service to determine which type of reports screen to show based on user's office
 */
export class ReportsRoutingService {
  private static cachedUserOffice: string | null = null;
  private static cachedIsDivisionUser: boolean | null = null;
  private static cacheTimestamp: Date | null = null;
  private static readonly CACHE_EXPIRY = 30 * 60 * 1000; // 30 minutes

  /**
   * Determines if the current user should see the comprehensive reports (Division users)
   * or the simple office-only reports (non-Division users)
   */
  static async shouldShowComprehensiveReports(): Promise<boolean> {
    try {
      // For debugging: always log cache status
      console.log('📋 ReportsRoutingService: shouldShowComprehensiveReports called');
      console.log('📋 ReportsRoutingService: Cache valid:', this.isCacheValid());
      console.log('📋 ReportsRoutingService: Cached isDivisionUser:', this.cachedIsDivisionUser);
      console.log('📋 ReportsRoutingService: Cached office:', this.cachedUserOffice);

      // Check cache first
      if (this.isCacheValid() && this.cachedIsDivisionUser !== null) {
        console.log('📋 ReportsRoutingService: Using cached result: isDivision=', this.cachedIsDivisionUser);
        return this.cachedIsDivisionUser;
      }

      console.log('📋 ReportsRoutingService: Cache invalid or empty, determining fresh report type...');

      const user = auth.currentUser;
      if (!user) {
        console.log('❌ ReportsRoutingService: No user logged in');
        return false;
      }

      console.log('📋 ReportsRoutingService: Current user UID:', user.uid);

      // Get user's office from Firebase
      const userDoc = await getDoc(doc(db, 'employees', user.uid));

      if (!userDoc.exists()) {
        console.log('❌ ReportsRoutingService: User document not found');
        return false;
      }

      const userData = userDoc.data();
      const officeName = userData?.officeName as string | undefined;

      if (!officeName || officeName.trim() === '') {
        console.log('❌ ReportsRoutingService: User office name not found');
        return false;
      }

      console.log('📋 ReportsRoutingService: User office:', `"${officeName}"`);

      // Check if office name ends with 'Division' (case-insensitive)
      const trimmed = officeName.trim();
      const lowercase = trimmed.toLowerCase();
      const isDivisionUser = lowercase.endsWith('division');

      console.log('📋 ReportsRoutingService: Office name:', `"${officeName}"`);
      console.log('📋 ReportsRoutingService: Trimmed:', `"${trimmed}"`);
      console.log('📋 ReportsRoutingService: Lowercase:', `"${lowercase}"`);
      console.log('📋 ReportsRoutingService: Ends with "division":', isDivisionUser);
      console.log('📋 ReportsRoutingService: Office type:', isDivisionUser ? 'DIVISION (Comprehensive Reports)' : 'OFFICE (Simple Reports)');

      // Cache the results
      this.cachedUserOffice = officeName;
      this.cachedIsDivisionUser = isDivisionUser;
      this.cacheTimestamp = new Date();

      console.log('📋 ReportsRoutingService: Cached results - Office:', this.cachedUserOffice);
      console.log('📋 ReportsRoutingService: Cached results - isDivision:', this.cachedIsDivisionUser);

      if (isDivisionUser) {
        console.log('✅ ReportsRoutingService: User is Division-level → Report Screen 1 (Comprehensive)');
      } else {
        console.log('✅ ReportsRoutingService: User is Office-level → Report Screen 2 (Table View Only)');
      }

      return isDivisionUser;

    } catch (error) {
      console.error('❌ ReportsRoutingService: Error determining report type:', error);
      return false; // Default to simple reports on error
    }
  }

  /**
   * Gets the current user's office name
   */
  static async getCurrentUserOfficeName(): Promise<string | null> {
    try {
      console.log('🔍 ReportsRoutingService: getCurrentUserOfficeName() called');

      // Use cached value if available and valid
      if (this.isCacheValid() && this.cachedUserOffice !== null) {
        console.log('📋 ReportsRoutingService: Using cached office name:', this.cachedUserOffice);
        return this.cachedUserOffice;
      }

      console.log('🔍 ReportsRoutingService: Cache invalid or empty, fetching fresh data');
      console.log('🔍 ReportsRoutingService: Cache valid:', this.isCacheValid());
      console.log('🔍 ReportsRoutingService: Cached office:', this.cachedUserOffice);

      const user = auth.currentUser;
      if (!user) {
        console.log('❌ ReportsRoutingService: No user logged in');
        return null;
      }

      console.log('🔍 ReportsRoutingService: Current user UID:', user.uid);
      console.log('🔍 ReportsRoutingService: Current user email:', user.email);

      console.log('🔍 ReportsRoutingService: Fetching user document from Firebase...');
      const userDoc = await getDoc(doc(db, 'employees', user.uid));

      if (!userDoc.exists()) {
        console.log('❌ ReportsRoutingService: User document does not exist in employees collection');
        console.log('🔍 ReportsRoutingService: Document path:', `employees/${user.uid}`);
        return null;
      }

      console.log('✅ ReportsRoutingService: User document found');
      const userData = userDoc.data();
      console.log('🔍 ReportsRoutingService: Raw user data:', userData);

      const officeName = userData?.officeName as string | undefined;
      console.log('🔍 ReportsRoutingService: Extracted officeName:', officeName);
      console.log('🔍 ReportsRoutingService: officeName type:', typeof officeName);
      console.log('🔍 ReportsRoutingService: officeName is null/undefined:', officeName == null);

      // Check all possible office-related fields
      console.log('🔍 ReportsRoutingService: All user data keys:', Object.keys(userData || {}));
      console.log('🔍 ReportsRoutingService: userData.officeName:', userData?.officeName);
      console.log('🔍 ReportsRoutingService: userData.office_name:', userData?.office_name);
      console.log('🔍 ReportsRoutingService: userData.office:', userData?.office);
      console.log('🔍 ReportsRoutingService: userData.Office:', userData?.Office);

      // Update cache
      this.cachedUserOffice = officeName || null;
      this.cacheTimestamp = new Date();

      console.log('🔍 ReportsRoutingService: Final office name result:', officeName || 'NULL');
      return officeName || null;

    } catch (error) {
      console.error('❌ ReportsRoutingService: Error getting user office:', error);
      console.error('❌ ReportsRoutingService: Error details:', error);
      return null;
    }
  }

  /**
   * Gets detailed information about the user's office type and access level
   */
  static async getUserOfficeInfo(): Promise<{
    officeName: string | null;
    isDivisionUser: boolean;
    accessLevel: string;
    reportType: string;
    description: string;
  }> {
    try {
      console.log('🔍 getUserOfficeInfo: Starting fresh analysis...');

      // Clear cache to ensure fresh data
      this.clearCache();

      const officeName = await this.getCurrentUserOfficeName();
      console.log('🔍 getUserOfficeInfo: Got office name:', officeName);

      // Use direct division logic test instead of cached shouldShowComprehensiveReports
      let isDivisionUser = false;
      if (officeName && officeName.trim() !== '') {
        isDivisionUser = officeName.trim().toLowerCase().endsWith('division');
        console.log('🔍 getUserOfficeInfo: Direct division check:', isDivisionUser);
        console.log('🔍 getUserOfficeInfo: Office trimmed lowercase:', `"${officeName.trim().toLowerCase()}"`);
        console.log('🔍 getUserOfficeInfo: Ends with "division":', isDivisionUser);
      }

      if (!officeName) {
        return {
          officeName: null,
          isDivisionUser: false,
          accessLevel: 'none',
          reportType: 'none',
          description: 'No office information available',
        };
      }

      let accessLevel: string;
      let reportType: string;
      let description: string;

      if (isDivisionUser) {
        accessLevel = 'division';
        reportType = 'comprehensive';
        description = 'Report Screen 1: Summary + Submissions + Table View tabs with multi-level office hierarchy data';
      } else {
        accessLevel = 'office';
        reportType = 'simple';
        description = 'Report Screen 2: Table View only with office-specific data';
      }

      console.log('🔍 getUserOfficeInfo: Final result:', {
        officeName,
        isDivisionUser,
        accessLevel,
        reportType
      });

      return {
        officeName,
        isDivisionUser,
        accessLevel,
        reportType,
        description,
      };

    } catch (error) {
      console.error('❌ ReportsRoutingService: Error getting office info:', error);
      return {
        officeName: null,
        isDivisionUser: false,
        accessLevel: 'error',
        reportType: 'simple',
        description: 'Error loading office information',
      };
    }
  }

  /**
   * Forces a fresh check bypassing all cache (for debugging)
   */
  static async forceRefreshDivisionStatus(): Promise<boolean> {
    console.log('🔄 ReportsRoutingService: Force refresh - bypassing all cache');

    // Clear cache completely
    this.clearCache();

    // Force fresh check
    const result = await this.shouldShowComprehensiveReports();

    console.log('🔄 ReportsRoutingService: Force refresh result:', result);
    return result;
  }

  /**
   * Clears the cache (useful for testing or when user data changes)
   */
  static clearCache(): void {
    this.cachedUserOffice = null;
    this.cachedIsDivisionUser = null;
    this.cacheTimestamp = null;
    console.log('🗑️ ReportsRoutingService: Cache cleared');
  }

  /**
   * Checks if the cached data is still valid
   */
  private static isCacheValid(): boolean {
    if (!this.cacheTimestamp) {
      return false;
    }
    
    const now = new Date();
    const difference = now.getTime() - this.cacheTimestamp.getTime();
    return difference < this.CACHE_EXPIRY;
  }

  /**
   * Gets a human-readable description of the user's report access
   */
  static async getAccessDescription(): Promise<string> {
    try {
      const officeInfo = await this.getUserOfficeInfo();

      if (!officeInfo.officeName) {
        return 'No office information available';
      }

      if (officeInfo.isDivisionUser) {
        return 'Division-level access: You can view Report Screen 1 with comprehensive reports including Summary, Submissions, and Table View tabs with multi-level office hierarchy data.';
      } else {
        return `Office-level access: You can view Report Screen 2 with Table View only containing data specific to your office (${officeInfo.officeName}).`;
      }

    } catch (error) {
      return 'Error determining access level';
    }
  }

  /**
   * Direct test of division logic without cache (for debugging)
   */
  static async testDivisionLogicDirect(): Promise<{
    officeName: string | null;
    trimmed: string | null;
    lowercase: string | null;
    endsWithDivision: boolean;
    shouldShowComprehensive: boolean;
  }> {
    try {
      console.log('🧪 === DIRECT DIVISION LOGIC TEST ===');

      // Get fresh data without cache
      const user = auth.currentUser;
      if (!user) {
        console.log('❌ Direct test: No user logged in');
        return {
          officeName: null,
          trimmed: null,
          lowercase: null,
          endsWithDivision: false,
          shouldShowComprehensive: false,
        };
      }

      const userDoc = await getDoc(doc(db, 'employees', user.uid));
      if (!userDoc.exists()) {
        console.log('❌ Direct test: User document not found');
        return {
          officeName: null,
          trimmed: null,
          lowercase: null,
          endsWithDivision: false,
          shouldShowComprehensive: false,
        };
      }

      const userData = userDoc.data();
      const officeName = userData?.officeName as string | undefined;

      console.log('🧪 Direct test: Raw office name:', officeName);

      if (!officeName) {
        console.log('❌ Direct test: Office name is null/undefined');
        return {
          officeName: null,
          trimmed: null,
          lowercase: null,
          endsWithDivision: false,
          shouldShowComprehensive: false,
        };
      }

      const trimmed = officeName.trim();
      const lowercase = trimmed.toLowerCase();
      const endsWithDivision = lowercase.endsWith('division');

      console.log('🧪 Direct test: Original:', `"${officeName}"`);
      console.log('🧪 Direct test: Trimmed:', `"${trimmed}"`);
      console.log('🧪 Direct test: Lowercase:', `"${lowercase}"`);
      console.log('🧪 Direct test: Ends with "division":', endsWithDivision);
      console.log('🧪 Direct test: Should show comprehensive:', endsWithDivision);

      console.log('🧪 === END DIRECT DIVISION LOGIC TEST ===');

      return {
        officeName,
        trimmed,
        lowercase,
        endsWithDivision,
        shouldShowComprehensive: endsWithDivision,
      };

    } catch (error) {
      console.error('❌ Direct test: Error:', error);
      return {
        officeName: null,
        trimmed: null,
        lowercase: null,
        endsWithDivision: false,
        shouldShowComprehensive: false,
      };
    }
  }

  /**
   * Logs detailed information about the user's report access (for debugging)
   */
  static async logUserAccessInfo(): Promise<void> {
    try {
      console.log('📋 === ReportsRoutingService: User Access Information ===');

      const officeInfo = await this.getUserOfficeInfo();

      console.log('📋 Office Name:', officeInfo.officeName);
      console.log('📋 Is Division User:', officeInfo.isDivisionUser);
      console.log('📋 Access Level:', officeInfo.accessLevel);
      console.log('📋 Report Type:', officeInfo.reportType);
      console.log('📋 Description:', officeInfo.description);

      const accessDescription = await this.getAccessDescription();
      console.log('📋 Access Description:', accessDescription);

      console.log('📋 === End User Access Information ===');

    } catch (error) {
      console.error('❌ ReportsRoutingService: Error logging access info:', error);
    }
  }
}

export default ReportsRoutingService;

File: lib/services/reports_routing_service.dart
import 'package:firebase_auth/firebase_auth.dart';
import 'package:cloud_firestore/cloud_firestore.dart';

/// Service to determine which type of reports screen to show based on user's office
class ReportsRoutingService {
  static final FirebaseAuth _auth = FirebaseAuth.instance;
  static final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  // Cache for user office type to avoid repeated Firebase calls
  static String? _cachedUserOffice;
  static bool? _cachedIsDivisionUser;
  static DateTime? _cacheTimestamp;
  static const Duration _cacheExpiry = Duration(minutes: 30);

  /// Determines if the current user should see the comprehensive reports (Division users)
  /// or the simple office-only reports (non-Division users)
  static Future<bool> shouldShowComprehensiveReports() async {
    try {
      // Check cache first
      if (_isCacheValid()) {
        print(
            '📋 ReportsRoutingService: Using cached result: isDivision=$_cachedIsDivisionUser');
        return _cachedIsDivisionUser ?? false;
      }

      print(
          '📋 ReportsRoutingService: Determining report type for current user...');
      print('📋 ReportsRoutingService: Cache valid: ${_isCacheValid()}');

      final user = _auth.currentUser;
      if (user == null) {
        print('❌ ReportsRoutingService: No user logged in');
        return false;
      }

      print('📋 ReportsRoutingService: Current user UID: ${user.uid}');

      // Get user's office from Firebase
      final userDoc =
          await _firestore.collection('employees').doc(user.uid).get();

      if (!userDoc.exists) {
        print('❌ ReportsRoutingService: User document not found');
        return false;
      }

      final userData = userDoc.data() as Map<String, dynamic>;
      final officeName = userData['officeName'] as String?;

      if (officeName == null || officeName.trim().isEmpty) {
        print('❌ ReportsRoutingService: User office name not found');
        return false;
      }

      print('📋 ReportsRoutingService: User office: "$officeName"');

      // Check if office name ends with 'Division' or contains 'Region' (case-insensitive)
      final trimmedLower = officeName.trim().toLowerCase();
      final isDivisionUser = trimmedLower.endsWith('division');
      final isRegionUser =
          trimmedLower.contains('region') && !trimmedLower.endsWith('division');
      final isComprehensiveUser = isDivisionUser || isRegionUser;

      print('📋 ReportsRoutingService: Office name: "$officeName"');
      print('📋 ReportsRoutingService: Trimmed lowercase: "$trimmedLower"');
      print('📋 ReportsRoutingService: Ends with "division": $isDivisionUser');
      print('📋 ReportsRoutingService: Contains "region": $isRegionUser');
      print(
          '📋 ReportsRoutingService: Comprehensive user (division OR region): $isComprehensiveUser');

      String officeType;
      if (isDivisionUser) {
        officeType = "DIVISION (Report Screen 1)";
      } else if (isRegionUser) {
        officeType = "REGION (Report Screen 1)";
      } else {
        officeType = "OFFICE (Report Screen 2)";
      }
      print('📋 ReportsRoutingService: Office type: $officeType');

      // Cache the results
      _cachedUserOffice = officeName;
      _cachedIsDivisionUser =
          isComprehensiveUser; // Cache comprehensive user flag
      _cacheTimestamp = DateTime.now();

      if (isComprehensiveUser) {
        if (isDivisionUser) {
          print(
              '✅ ReportsRoutingService: User is Division-level → Report Screen 1 (Comprehensive)');
        } else if (isRegionUser) {
          print(
              '✅ ReportsRoutingService: User is Region-level → Report Screen 1 (Comprehensive)');
        }
      } else {
        print(
            '✅ ReportsRoutingService: User is Office-level → Report Screen 2 (Table View Only)');
      }

      return isComprehensiveUser;
    } catch (error) {
      print('❌ ReportsRoutingService: Error determining report type: $error');
      return false; // Default to simple reports on error
    }
  }

  /// Gets the current user's office name
  static Future<String?> getCurrentUserOfficeName() async {
    try {
      // Use cached value if available and valid
      if (_isCacheValid() && _cachedUserOffice != null) {
        return _cachedUserOffice;
      }

      final user = _auth.currentUser;
      if (user == null) {
        return null;
      }

      final userDoc =
          await _firestore.collection('employees').doc(user.uid).get();

      if (!userDoc.exists) {
        return null;
      }

      final userData = userDoc.data() as Map<String, dynamic>;
      final officeName = userData['officeName'] as String?;

      // Update cache
      _cachedUserOffice = officeName;
      _cacheTimestamp = DateTime.now();

      return officeName;
    } catch (error) {
      print('❌ ReportsRoutingService: Error getting user office: $error');
      return null;
    }
  }

  /// Gets detailed information about the user's office type and access level
  static Future<Map<String, dynamic>> getUserOfficeInfo() async {
    try {
      print('🔍 getUserOfficeInfo: Starting fresh analysis...');

      // Clear cache to ensure fresh data
      clearCache();

      final officeName = await getCurrentUserOfficeName();
      print('🔍 getUserOfficeInfo: Got office name: $officeName');

      // Use direct division/region logic test instead of cached shouldShowComprehensiveReports
      bool isDivisionUser = false;
      bool isRegionUser = false;
      bool isComprehensiveUser = false;

      if (officeName != null && officeName.trim().isNotEmpty) {
        final trimmedLower = officeName.trim().toLowerCase();
        isDivisionUser = trimmedLower.endsWith('division');
        isRegionUser = trimmedLower.contains('region') &&
            !trimmedLower.endsWith('division');
        isComprehensiveUser = isDivisionUser || isRegionUser;

        print('🔍 getUserOfficeInfo: Direct division check: $isDivisionUser');
        print('🔍 getUserOfficeInfo: Direct region check: $isRegionUser');
        print(
            '🔍 getUserOfficeInfo: Comprehensive user (division OR region): $isComprehensiveUser');
        print(
            '🔍 getUserOfficeInfo: Office trimmed lowercase: "$trimmedLower"');
        print('🔍 getUserOfficeInfo: Ends with "division": $isDivisionUser');
        print('🔍 getUserOfficeInfo: Contains "region": $isRegionUser');
      }

      if (officeName == null) {
        return {
          'officeName': null,
          'isDivisionUser': false,
          'accessLevel': 'none',
          'reportType': 'none',
          'description': 'No office information available',
        };
      }

      String accessLevel;
      String reportType;
      String description;

      if (isComprehensiveUser) {
        if (isDivisionUser) {
          accessLevel = 'division';
        } else if (isRegionUser) {
          accessLevel = 'region';
        } else {
          accessLevel = 'comprehensive';
        }
        reportType = 'comprehensive';
        description =
            'Report Screen 1: Summary + Submissions + Table View tabs with multi-level office hierarchy data';
      } else {
        accessLevel = 'office';
        reportType = 'simple';
        description =
            'Report Screen 2: Table View only with office-specific data';
      }

      print(
          '🔍 getUserOfficeInfo: Final result: {officeName: $officeName, isDivisionUser: $isDivisionUser, accessLevel: $accessLevel, reportType: $reportType}');

      return {
        'officeName': officeName,
        'isDivisionUser':
            isComprehensiveUser, // Changed to use comprehensive user flag
        'isRegionUser': isRegionUser,
        'isComprehensiveUser': isComprehensiveUser,
        'accessLevel': accessLevel,
        'reportType': reportType,
        'description': description,
      };
    } catch (error) {
      print('❌ ReportsRoutingService: Error getting office info: $error');
      return {
        'officeName': null,
        'isDivisionUser': false,
        'accessLevel': 'error',
        'reportType': 'simple',
        'description': 'Error loading office information',
      };
    }
  }

  /// Clears the cache (useful for testing or when user data changes)
  static void clearCache() {
    _cachedUserOffice = null;
    _cachedIsDivisionUser = null;
    _cacheTimestamp = null;
    print('🗑️ ReportsRoutingService: Cache cleared');
  }

  /// Checks if the cached data is still valid
  static bool _isCacheValid() {
    if (_cacheTimestamp == null) {
      return false;
    }

    final now = DateTime.now();
    final difference = now.difference(_cacheTimestamp!);
    return difference < _cacheExpiry;
  }

  /// Gets a human-readable description of the user's report access
  static Future<String> getAccessDescription() async {
    try {
      final officeInfo = await getUserOfficeInfo();
      final officeName = officeInfo['officeName'] as String?;
      final isDivisionUser = officeInfo['isDivisionUser'] as bool;

      if (officeName == null) {
        return 'No office information available';
      }

      if (isDivisionUser) {
        return 'Division-level access: You can view Report Screen 1 with comprehensive reports including Summary, Submissions, and Table View tabs with multi-level office hierarchy data.';
      } else {
        return 'Office-level access: You can view Report Screen 2 with Table View only containing data specific to your office ($officeName).';
      }
    } catch (error) {
      return 'Error determining access level';
    }
  }

  /// Logs detailed information about the user's report access (for debugging)
  static Future<void> logUserAccessInfo() async {
    try {
      print('📋 === ReportsRoutingService: User Access Information ===');

      final officeInfo = await getUserOfficeInfo();

      print('📋 Office Name: ${officeInfo['officeName']}');
      print('📋 Is Division User: ${officeInfo['isDivisionUser']}');
      print('📋 Access Level: ${officeInfo['accessLevel']}');
      print('📋 Report Type: ${officeInfo['reportType']}');
      print('📋 Description: ${officeInfo['description']}');

      final accessDescription = await getAccessDescription();
      print('📋 Access Description: $accessDescription');

      print('📋 === End User Access Information ===');
    } catch (error) {
      print('❌ ReportsRoutingService: Error logging access info: $error');
    }
  }
}
